docs/design-proposals/node/Unified-Vertical-Scaling-for-Container-and-VM-Pods.md

```diff
@@ -319,3 +341,129 @@ Other components:
 * check how the change of meaning of resource requests influence other
   Kubernetes components.
 
+### Changes in Arktos-vm-runtime-service (CRI for VM workload types)
+
+#### Specific for VM workload
+
+1. Hotplug(hotUnplug) vs Memory Ballooning driver for Memory resizing
+
+   Currently per our knowledge these are two technologies that supports live memory resizing for VMs. Both have been there for
+   Linux for years, and both has pros and cons in supporting memory resizing for VMs. There are also some proposals to combine
+   them to better support VM resizing for Linux VMs.
+
+   In Arktos VM runtime, it is proposed to use hotplug(hotUnplug) in current release for the following reasons:
+
+    1. Numa support
+    2. Less memory fragmentation
+    3. Better control from VM spec patching
+
+2. VM memory live update
+
+    Both attach and detach device operations are done via the AttachDeviceFlag and DetachDeviceFlag API, with flag to
+    affect the running VM and persist the changes to the domain configuration for next reboot as well.
+       
+       AttachDeviceFlags(memoryDeviceDefinition, libvirt.DOMAIN_DEVICE_MODIFY_CONFIG|libvirt.DOMAIN_DEVICE_MODIFY_LIVE)
+       DetachDeviceFlags(memoryDeviceDefinition, libvirt.DOMAIN_DEVICE_MODIFY_CONFIG|libvirt.DOMAIN_DEVICE_MODIFY_LIVE)
+   
+3. Handle the async operation for memory unplug
+
+   Add a handler for the VIR_DOMAIN_EVENT_ID_DEVICE_REMOVED (future works). Current design returns in the UpdateContainerResource()
+   once after the DetachDeviceFlag() returns success without checking if the memory device is removed at the VM configuration in
+   the guest. It is done such way mainly for not holding the UpdateContainerResource() API for unexpected longer time due to
+   memory unplug.
+   
+4. Some limitations/constraints for VM workload types
+
+    1. Memory device size: 
+       Currently it is single size, with 128MiB each. i.e. if a VM patch increases the memory as 1Gi, 8 memory devices 
+       will be attached to the VM. The operations will still in one UpdateContainerResource() call though. i.e. Arktos VM 
+       runtime service will handle the attach operation internally with the calculation of memory increment in the patch.
+       
+    2. Max memory scale-up value
+       Currently a VM memory slots are set to 16 in Arktos VM runtime, so 16 * 128 Mi, which is 2Gi max scale-up value.
```

Comment:
Max 2G memory? Seriously? :). It seems too small

Reply:
this is for the increase for the current implementation. it can be adjusted but i am not sure how much we expect it to be.

Reply:
A memoryDevice simulates a memory chip in real world.  128M is too small for today's world.  We can change it to a larger value.

Reply:
per the discussion today. will update the design with a "multiple devices size support" approach.